### ISSUE TITLE
Unspecified behaviour with explicit writes to minstret on RV32

### ISSUE THREAD

**Issue by Timmmm** [riscv/riscv-isa-manual#1255]
On RV32 if `minstret` the lower 32 bits of `instret` is 0xFFFFFFFF, and `minstreth` (the upper 32 bits) is 0. And then you do

```
csrw minstret, zero
```

What value would you expect `minstreth` to get? 0 or 1?

The two possibilities are:

0: You have explicitly written to `instret`, therefore the automatic increment is suppressed. As sort of indicated by this text:

> Some CSRs, such as the instructions-retired counter, instret, may be modified as side effects of instruction execution. In these cases, if a CSR access instruction reads a CSR, it reads the value prior to the execution of the instruction. If a CSR access instruction writes such a CSR, the write is done instead of the increment. In particular, a value written to instret by one instruction will be the value read by the following instruction.
> ...
> Furthermore, an explicit CSR read returns the CSR state before the execution of the instruction, while an explicit CSR write suppresses and overrides any implicit writes or modifications to the same CSR by the same instruction.

1: The instruction completes as usual, incrementing the counter, and *then* the write takes effect. As sort of indicated by this text:

> Any CSR write takes effect after the writing instruction has otherwise completed.

I suspect the answer is 0, but it's not really clear because it talks about `instret` (which isn't actually the name of a CSR), and "the same CSR" - but is `minstret` the same CSR as `minstreth` on RV32?

**Comment by gfavor** [comment]
The arch spec says:

Some CSRs, such as the instructions-retired counter, instret, may be modified as side effects of
instruction execution. In these cases, if a CSR access instruction reads a CSR, it reads the value
prior to the execution of the instruction. If a CSR access instruction writes such a CSR, the write
is done instead of the increment. In particular, a value written to instret by one instruction will
be the value read by the following instruction.

**Comment by Timmmm** [comment]
Yes I quoted that section. It's very clear for RV64 where `instret` (which isn't really a CSR) is exactly equal to `minstret`. But on RV32 it is split between `minstret` and `minstreth`, so the intended behaviour is not fully specified.

**Comment by Timmmm** [comment]
I looked into what Spike does. The code goes like this:

```
// Write to minstreth. `orig` is wide_counter_csr_t.
bool rv32_high_csr_t::unlogged_write(const reg_t val) noexcept {
  return orig->unlogged_write((orig->written_value() << 32 >> 32) | ((val & 0xffffffffU) << 32));
}

...

// Write to 64-bit instret.
bool wide_counter_csr_t::unlogged_write(const reg_t val) noexcept {
  this->val = val;
  // The ISA mandates that if an instruction writes instret, the write
  // takes precedence over the increment to instret.  However, Spike
  // unconditionally increments instret after executing an instruction.
  // Correct for this artifact by decrementing instret here.
  // Ensure that Smctrpmf hasn't disabled counting.
  if (is_counting_enabled()) {
    this->val--;
  }
  return true;
}

...

// Increment after instruction retired. `howmuch` is 0 or 1.
void wide_counter_csr_t::bump(const reg_t howmuch) noexcept {
  if (is_counting_enabled()) {
    val += howmuch;  // to keep log reasonable size, don't log every bump
  }
  // Clear cached value
  config_csr->reset_prev();
}
```

In other words, any time you write to 32-bit `minstret` or `minstreth`, it applies the write immediately and then decrements 64-bit `instret`. Then it always increments it after the instruction has retired.

This made me think it would be doing option 0. If you write to either half of 64-bit `instret` then the other half doesn't change. However I must be missing something from reading the code because the actual behaviour is option 1, as shown in this instruction trace:

```
core   0: 0x80000048 (0xb8201073) csrw    minstreth, zero           # Write 0x00000000 to minstreth.
core   0: 3 0x80000048 (0xb8201073) c2946_minstreth 0x00000000
core   0: 0x8000004c (0xb0229073) csrw    minstret, t0              # Write 0xFFFFFFFF to minstret.
core   0: 3 0x8000004c (0xb0229073) c2818_minstret 0xffffffff       # `instret` must now be 0x00000000FFFFFFFF with either option.
core   0: 0x80000050 (0xb8215073) csrwi   minstreth, 2              # Write 2 to minstreth. After this it is either
core   0: 3 0x80000050 (0xb8215073) c2946_minstreth 0x00000002      # 0x00000002FFFFFFFF for option 0 (increment is suppressed) or 0x0000000200000000 for option 1 (increment before write).
core   0: 0x80000054 (0xb0202373) csrr    t1, minstret              # Read minstret. It is 0 => option 0.
core   0: 3 0x80000054 (0xb0202373) x6  0x00000000
```

The Sail model currently does option 1, but I have a patch to change that. It does complicate the code a bit though.

**Comment by Timmmm** [comment]
Ah the bit of the code I missed is that `written_value()` returns the value after the `bump()`:

```
reg_t wide_counter_csr_t::written_value() const noexcept {
  // Re-adjust for upcoming bump()
  return this->val + 1;
}
```

This actually seems like a bug in Spike. If `mcountinhibit` disables `minstret` increment, and `instret` is `0x00000000FFFFFFFF` and you write 2 to `minstreth` then Spike will say `instret` is now `0x0000000200000000` which is clearly wrong. But aside from that it seems pretty clear that Spike is going for option 1 (increment then write).

I'll maybe check some open source cores to see what they do too.

**Comment by Timmmm** [comment]
We'd really appreciate some clarify on this, so I've made it really easy for you - two pull requests that specify the two options. You just have to pick one!

* Spike implements #1948
* Valtrix STING expects #1949 
* The Sail model originally implemented #1949 but I changed it to #1948 recently.

I'm kind of leaning towards #1949. It seems kind of more useful for software. But it also relies more heavily on the idea of multiple CSRs being one CSR... which is kind of woolly in the spec at the moment.

Anyway either option would be better than nothing!

**Comment by aswaterman** [comment]
I do agree Spike needs to be fixed, though.  I will work on that.

**Comment by aswaterman** [comment]
(Asking @gfavor for a second opinion)

**Comment by gfavor** [comment]
Setting aside the Spike bug, effectively I agree with Andrew.  Regarding #1948, I woudl also say that its wordsing conflates the CSRs for accessing portions of underlying architectural state, i.e. there is the underlying 64-bit register and then (in RV32) there are two CSRs for accessing one half or the other of the register.  And, as Andrew noted, whether one is reading or writing just part of the register, one is "accessing the register".  In the case of writing, this is effectively an atomic RMW of the register.

**Comment by Timmmm** [comment]
> It's hard to imagine we'd be discussing this topic if not for that fact.

Maybe, but shouldn't the behaviour be clear from the spec and not require consulting Spike?

>  its wordsing conflates the CSRs for accessing portions of underlying architectural state, i.e. there is the underlying 64-bit register and then (in RV32) there are two CSRs for accessing one half or the other of the register

The existing spec already conflates those two things - the underlying 64-bit register is "a CSR" and the XLEN views of it are also CSRs:

> The minstret CSR counts the number of instructions the hart has retired. The mcycle and minstret registers have 64-bit precision on all RV32 and RV64 harts.
> ...
> When XLEN=32, reads of the mcycle, minstret, mhpmcountern, and mhpmeventn CSRs return bits 31-0 of the corresponding register, and writes change only bits 31-0; reads of the mcycleh, minstreth, mhpmcounternh, and mhpmeventnh CSRs return bits 63-32 of the corresponding register, and writes change only bits 63-32.

This existing conflation is one of the reasons this needs clarifying!

**Comment by gfavor** [comment]
The first paragraph of Priv 3.1.10 describes the minstret CSR as counting instruction retires, and states that the minstret register has 64-bit precision.  Then the 4th paragraph says "When XLEN=32, reads of the minstret CSR returns bits 31-0 of the
corresponding register, and writes change only bits 31-0; reads of the minstreth CSR returns bits 63-32 of the corresponding register, and writes change only bits 63-32. "

In short, the spec describes a 64-bit-wide "minstret" register, a "minstret" CSR that accesses the lower 32 or 64 bits of the register depending on RV32/RV64, and a "minstreth" CSR in RV32 that accesses the high 32 bits of the minstret register.  Personally I would say that the spec is pretty clear.  One could argue that the register and the CSR names should have been different from each other (and this would need to be repeated across the architecture for other CSRs/registers to be consistent), but the consistent pattern across the architecture currently is to use the same name for both the CSR and the underlying register.

**Comment by allenjbaum** [comment]
Sorry for the noise; I can't find the beginning of this thread, so don't
recall what the answer was.
For an RV32:
if minstret is written on a cycle that would have generated a carry from
bit 31 to bit32,
 - does the upper half increment, or is the carry suppressed?
The wording " and writes change only bits 31-0" could imply that the carry
is suppressed, except the write isn't changing of the upper half, it is the
side effect of the incrementer that does that. Similarly:
if minstreth is written on a cycle that would have generated a carry from
bit 31 to bit32,
 - does the upper half increment (as if the write, happened before the
increment) or is it suppressed (as if write happened after the increment)

**Comment by gfavor** [comment]
Think of accesses to any CSR, whether to all of it or just to part of it, as an atomic access - a RMW in the case of a partial write.  (In fact all CSR instructions are architecturally defined to atomically read-modify-write a CSR.)

Then the only question in the case of minstret is whether the increment of the register by the retire of the CSR access instruction semantically occurs before or after the access.  And section 6.1.1 of the Unpriv spec specifies this behavior wrt read and write accesses.

**Comment by Timmmm** [comment]
>

**Comment by Timmmm** [comment]
Oops didn't meant to close this. Can someone reopen it?

> In short, the spec describes a 64-bit-wide "minstret" register

Yes but it also describes these 64-bit counter registers as CSRs.

> The minstret CSR counts the number of instructions the hart has retired.

If you're really saying that that isn't a CSR then at least that should say "The minstret register counts..."

**Comment by wmat** [comment]
Reopened @Timmmm

**Comment by gfavor** [comment]
As I mentioned in my last post, I effectively don't disagree that the spec talks about a minstret register and a minstret CSR and that they are 64 bits wide.  I won't repeat the whole post, but just this excerpt:

> the spec describes a 64-bit-wide "minstret" register, a "minstret" CSR that accesses the lower 32 or 64 bits of the register depending on RV32/RV64, and a "minstreth" CSR in RV32 that accesses the high 32 bits of the minstret register.

Practically speaking the CSR and the register are both describing the same counter that counts instruction retires.  The 'CSR' is just the mapping into CSR address space of the register for accessing purposes.

While one can argue about the dual use of the label 'minstret', this is a very common pattern across the architecture and has generally not been a source of confusion.

**Comment by Timmmm** [comment]
So this is wrong then:

> The minstret CSR counts the number of instructions the hart has retired.

It should say "the minstret register..."

**Comment by gfavor** [comment]
The register and the CSR are both describing the same counter that counts instruction retires.  One can argue over whether one should say that the register counts instret's, or that the CSR counts instret's, but for the purposes of conveying the architectural concept they are same due to the 1:1 relation between the 'register' and the 'CSR".  (The CSR perspective is what matters to a software person; the register perspective is what matters to a hardware person.  Although that's just getting picky over the wording.)

At the end of the day, the spec could have been written differently, but it is what it is - and it pretty clearly expresses the 1:1 correspondence between the two terms, i.e. they essentially are synonyms for the same thing that counts instret's.

**Comment by Timmmm** [comment]
Nobody else I've discussed this with thought the spec gave a clear answer, so I think observationally it doesn't clearly express that `minstret` and `minstreth` are "the same CSR".

**Comment by gfavor** [comment]
The spec says:

> When XLEN=32, reads of the minstret CSR returns bits 31-0 of the corresponding register, and writes change only bits 31-0; reads of the minstreth CSR returns bits 63-32 of the corresponding register, and writes change only bits 63-32.

Whereas, when XLEN=64, minstret CSR accesses read/write the full 64 bits of the register.

So minstret and minsreth are not literally the same CSR (they have different CSR numbers), but they are, if you will, aliases for accessing all or parts of the underlying 64 bits of state.

Also note that this is the reason why the text talks in terms of both the minstret CSR and the minstret register.  The register is always 64-bits wide, while the minstret and minstreth CSRs are XLEN-wide.  From a hardware perspective the register is what counts instruction retires and the CSRs are the address mapping for accessing the counter.  While from a software perspective the minstret CSR counts instruction retires (and the register is an implementation detail transparent to software).

Also note that the pattern of a *h CSR existing only in RV32 for accesssing the upper half of a 64-bit register for when the * CSR can only access the low half of the register in RV32, is very common and consistent across the architecture.  While at the same time the base or initial RV64-oriented description of a 64-bit CSR/register treats the two as synonymous since they are.

**Comment by kasanovic** [comment]
There are underlying CSRs holding architectural state, and there are CSR instructions that access those CSRs using CSR addresses.  The confusion here is between CSR addresses used in CSR instructions and the CSRs themselves - there is not a 1-1 mapping.  It is not uncommon in the architecture for execution of a CSR instruction at one address to affect values seen by a CSR instruction accessing a different address (as in shadowed views).  It is not practical to try to separate these two concepts, in particular when the same CSR state is accessed from two privileged modes, where it would break abstractions to describe the underlying CSR that has fields only accessible to a higher mode, or where writes to a lower-privileged mode affect state seen by a higher-privilege mode (and where some implementations might not have the higher-privileged mode at all).

**Comment by Timmmm** [comment]
That makes sense, but it's still true that the ISA manual confuses these. For example by that logic `minstreth` is not a CSR; it is a CSR address. However it is described as a CSR in the ISA manual:

> ...reads of the mcycleh, minstreth, mhpmcounternh, and mhpmeventnh CSRs...

> Analogously, when XLEN=32, the cycleh, instreth and hpmcounternh CSRs are read-only shadows of mcycleh, minstreth and mhpmcounternh, respectively. 

Also the CSR table is described:

> Table 4 - Table 8 list the CSRs that have currently been allocated CSR addresses.

> It is not practical to try to separate these two concepts

Yeah probably not but then doesn't that mean that you can't just say "it's obvious already"? There are two different concepts and the spec doesn't properly separate them, so when it says "the same CSR" that surely needs some clarification as to which concept it is talking about?

Anyway I'll close this since there doesn't seem to be much appetite for fixing it and at least we have a riscv-test and this spec-by-github-issue! :-D

**Comment by radimkrcmar** [comment]
Clarity suffers when using the same word with different meanings or different words with the same meaning.

The manual also uses "instret counter" and all gets further tangled since an implementation can represent the architectural state in any way it wants, as long as the observable behavior (the CSR interface) is correct.

It does add a poetic depth to the manual, though. :)

**Comment by kasanovic** [comment]
The manual does need extra clarification around CSR accesses, but this is a more global issue than just minstreth.  I'll create separate issue to track clarification around CSRs/CSR accesses in the Zicsr chapter.
